{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Blasting known HCN genes against the white clover reference genome\n",
    "\n",
    "So far, we have not been able to differentiate homzygous dominant (e.g., AcAc) white clover individuals from heterozygous (e.g., Acac) individuals using PCR since the dominance of the markers and the multiple deletions at both loci result in preferential amplification of the smaller deletion, masking the genotype on the other homologous chromosome. We therefore need to design a PCR primer that spans the entire deletion boundary, which would allow us to resolve homozygous and heterozygous individuals on a gel based on the number and size of bands.\n",
    "\n",
    "To facilitate this, we can blast the sequences of all known HCN genes against the recently published white clover genome (Griffiths _et al._ 2019, Plant Cell). Contigs with strong matches to genes can be retrieved and manually examined for unique sequence upstream of the deletion boundary from which we can design PCR primers.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import os\n",
    "import pandas as pd\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def exit_status(exit_code):\n",
    "    \"\"\"Prints success/fail of subprocess based on exit code\n",
    " \n",
    "    Args:\n",
    "        exit_code: ('int'): Exit code of subprocess\n",
    " \n",
    "    Returns:\n",
    "        None: Print to stdout.\n",
    "    \"\"\"\n",
    "    if exit_code == 0:\n",
    "        print(\"\\nSubprocess completed successfully \\n\")\n",
    "    else:\n",
    "        print(\"\\nThere was an error with the subprocess.\\n Double check the command and try again.\")\n",
    "\n",
    "        \n",
    "def makebBlastDatabase(fasta_in):\n",
    "    \"\"\"Creates BLAST database from fasta file.\n",
    " \n",
    "    Args:\n",
    "        fasta_in ('str'): Full path to input FASTA file from which database should be created\n",
    " \n",
    "    Returns\"\n",
    "        None: Writes BLAST database to disk\n",
    "    \"\"\" \n",
    "    # Name of database\n",
    "    db_out = fasta_in + '.db'\n",
    "  \n",
    "    # Command to make BLAST database\n",
    "    MakeBlastDb_Command = \"makeblastdb -in \" + fasta_in + \" -out \" + db_out + \" -parse_seqids\" + \" -dbtype nucl\"\n",
    "\n",
    "    # Check whether subprocess successfully completed\n",
    "    exit_code = subprocess.call(MakeBlastDb_Command, shell=True)\n",
    "    exit_status(exit_code)\n",
    "    \n",
    "    \n",
    "def blast_sequences(fasta_in, db, outfile):\n",
    "    \"\"\"BLAST's a set of sequences against BLAST database\n",
    " \n",
    "    Args:\n",
    "        fasta_in ('str'): Full path to input FASTA file from which database should be created\n",
    "        db ('str'): Full path to input BLAST database\n",
    "        outfile ('str'): Name of output file\n",
    " \n",
    "    Returns\"\n",
    "        None: Writes BLAST results with header to disk\n",
    "    \"\"\" \n",
    "    full_fasta_in_path = os.path.abspath(fasta_in)\n",
    "    outpath = '/' + '/'.join(full_fasta_in_path.split('/')[1:-1]) + 'HCN_genes_BLAST/' + outfile\n",
    "    \n",
    "    print(full_fasta_in_path)\n",
    "    print(outpath)\n",
    "    print(db)\n",
    "    \n",
    "#     Blast_Command = \"blastn -query {0} -db {1} -out {2} -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend qlen sstart send slen evalue bitscore qcovs qcovhsp' -num_threads 2 -evalue 1e-10 -max_target_seqs 1 -max_hsps 1\".format(fasta_in, db, outpath)\n",
    "    Blast_Command = \"blastn -query {0} -db {1} -out {2} -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend qlen sstart send slen evalue bitscore qcovs qcovhsp' -num_threads 2 -evalue 1e-10 -max_hsps 1 -max_target_seqs 1\".format(fasta_in, db, outpath)\n",
    "    print(Blast_Command)\n",
    "    exit_code = subprocess.call(Blast_Command, shell=True)\n",
    "    exit_status(exit_code)\n",
    "    \n",
    "    columns = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch', 'gapopen', 'qstart', 'qend', 'qlen', 'sstart', 'send', 'slen', 'evalue', 'bitscore', 'qcovs', 'qcovhsp']\n",
    "    blast_results = pd.read_table(outpath, sep = '\\t', names = columns)\n",
    "    \n",
    "    blast_results.to_csv(outpath, sep = '\\t', index = False)\n",
    "    \n",
    "\n",
    "def retrieve_contigs(blast_results, genome_fasta, outfile):\n",
    "    '''Writes FASTA with sequences of contigs that match HCN genes\n",
    "    \n",
    "    Args:\n",
    "        blast_results ('str'): Full path to text file with results from BLAST\n",
    "        genome_fasta ('str'): Full path to genome FASTA file with contigs to extract\n",
    "        outfile ('str'): Name of output file\n",
    "        \n",
    "    Returns:\n",
    "        None: Writes FASTA to disk\n",
    "    '''\n",
    "    \n",
    "    blast_df = pd.read_table(blast_results, sep = '\\t')\n",
    "    \n",
    "    sseqids = set(blast_df['sseqid'].tolist())\n",
    "    \n",
    "    genome_scaffolds = SeqIO.to_dict(SeqIO.parse(genome_fasta, 'fasta'))\n",
    "    \n",
    "    outpath = '/' + '/'.join(blast_results.split('/')[1:-1]) + '/' + outfile\n",
    "#     print(outpath)\n",
    "\n",
    "    with open(outpath, 'w') as f:\n",
    "        \n",
    "        for scaffold in sseqids:\n",
    "            description = genome_scaffolds[scaffold].description\n",
    "            sequence = genome_scaffolds[scaffold].seq\n",
    "            \n",
    "            f.write('>{0}\\n{1}\\n'.format(description, sequence))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1: Make BLAST database from genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Subprocess completed successfully \n",
      "\n"
     ]
    }
   ],
   "source": [
    "fasta_in = '/scratch/research/references/trifolium/repens/griffiths_2019_Plant-cell/Griffiths_et_al_genome_AllSeqs.fasta'\n",
    "makebBlastDatabase(fasta_in)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 2: BLAST sequences against database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/scratch/research/projects/trifolium/HCN_genes_in_genome/T-repens_HCN_genes.fasta\n",
      "/scratch/research/projects/trifolium/HCN_genes_in_genome/HCN_griffithsGenome_Blast.txt\n",
      "/scratch/research/references/trifolium/repens/griffiths_2019_Plant-cell/Griffiths_et_al_genome_AllSeqs.fasta.db\n",
      "blastn -query ../T-repens_HCN_genes.fasta -db /scratch/research/references/trifolium/repens/griffiths_2019_Plant-cell/Griffiths_et_al_genome_AllSeqs.fasta.db -out /scratch/research/projects/trifolium/HCN_genes_in_genome/HCN_griffithsGenome_Blast.txt -outfmt '6 qseqid sseqid pident length mismatch gapopen qstart qend qlen sstart send slen evalue bitscore qcovs qcovhsp' -num_threads 2 -evalue 1e-10 -max_hsps 1 -max_target_seqs 1\n",
      "\n",
      "Subprocess completed successfully \n",
      "\n"
     ]
    }
   ],
   "source": [
    "fasta_in = '../T-repens_HCN_genes.fasta'\n",
    "db = '/scratch/research/references/trifolium/repens/griffiths_2019_Plant-cell/Griffiths_et_al_genome_AllSeqs.fasta.db'\n",
    "outfile = 'HCN_griffithsGenome_Blast.txt'\n",
    "blast_sequences(fasta_in, db, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>qseqid</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>pident</th>\n",
       "      <th>length</th>\n",
       "      <th>mismatch</th>\n",
       "      <th>gapopen</th>\n",
       "      <th>qstart</th>\n",
       "      <th>qend</th>\n",
       "      <th>qlen</th>\n",
       "      <th>sstart</th>\n",
       "      <th>send</th>\n",
       "      <th>slen</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>qcovhsp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KJ467261.1</td>\n",
       "      <td>VCDJ01010665.1</td>\n",
       "      <td>99.909</td>\n",
       "      <td>1101</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1101</td>\n",
       "      <td>1101</td>\n",
       "      <td>19582530</td>\n",
       "      <td>19583630</td>\n",
       "      <td>54415314</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2028</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KJ467326.1</td>\n",
       "      <td>VCDJ01010665.1</td>\n",
       "      <td>97.737</td>\n",
       "      <td>1679</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1679</td>\n",
       "      <td>1679</td>\n",
       "      <td>19570565</td>\n",
       "      <td>19572208</td>\n",
       "      <td>54415314</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2857</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JQ920491.1</td>\n",
       "      <td>VCDJ01010665.1</td>\n",
       "      <td>95.318</td>\n",
       "      <td>1132</td>\n",
       "      <td>18</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>1128</td>\n",
       "      <td>1128</td>\n",
       "      <td>19582530</td>\n",
       "      <td>19583630</td>\n",
       "      <td>54415314</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1764</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JQ920574.1</td>\n",
       "      <td>VCDJ01010665.1</td>\n",
       "      <td>98.889</td>\n",
       "      <td>3420</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3392</td>\n",
       "      <td>3392</td>\n",
       "      <td>19567103</td>\n",
       "      <td>19570520</td>\n",
       "      <td>54415314</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6080</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JQ920575.1</td>\n",
       "      <td>VCDJ01010665.1</td>\n",
       "      <td>97.664</td>\n",
       "      <td>2226</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>1789</td>\n",
       "      <td>4014</td>\n",
       "      <td>7419</td>\n",
       "      <td>19581984</td>\n",
       "      <td>19584170</td>\n",
       "      <td>54415314</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3786</td>\n",
       "      <td>30</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>EF990447.1</td>\n",
       "      <td>VCDJ01010675.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>2225</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>756</td>\n",
       "      <td>2980</td>\n",
       "      <td>3839</td>\n",
       "      <td>30228279</td>\n",
       "      <td>30230503</td>\n",
       "      <td>51265060</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4109</td>\n",
       "      <td>58</td>\n",
       "      <td>58</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>JQ920572.1</td>\n",
       "      <td>VCDJ01010675.1</td>\n",
       "      <td>98.425</td>\n",
       "      <td>2920</td>\n",
       "      <td>25</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2901</td>\n",
       "      <td>2901</td>\n",
       "      <td>30224409</td>\n",
       "      <td>30227326</td>\n",
       "      <td>51265060</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5118</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>JQ920573.1</td>\n",
       "      <td>VCDJ01010675.1</td>\n",
       "      <td>99.604</td>\n",
       "      <td>4541</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>4072</td>\n",
       "      <td>8611</td>\n",
       "      <td>13521</td>\n",
       "      <td>30218214</td>\n",
       "      <td>30222753</td>\n",
       "      <td>51265060</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8285</td>\n",
       "      <td>34</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>JQ920570.1</td>\n",
       "      <td>VCDJ01010674.1</td>\n",
       "      <td>99.515</td>\n",
       "      <td>824</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>824</td>\n",
       "      <td>824</td>\n",
       "      <td>11421697</td>\n",
       "      <td>11422520</td>\n",
       "      <td>65212296</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1500</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>MH059954.1</td>\n",
       "      <td>VCDJ01010672.1</td>\n",
       "      <td>99.541</td>\n",
       "      <td>19606</td>\n",
       "      <td>62</td>\n",
       "      <td>7</td>\n",
       "      <td>59707</td>\n",
       "      <td>79297</td>\n",
       "      <td>197905</td>\n",
       "      <td>15981373</td>\n",
       "      <td>15961781</td>\n",
       "      <td>62546397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35682</td>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       qseqid          sseqid   pident  length  mismatch  gapopen  qstart  \\\n",
       "0  KJ467261.1  VCDJ01010665.1   99.909    1101         1        0       1   \n",
       "1  KJ467326.1  VCDJ01010665.1   97.737    1679         3        3       1   \n",
       "2  JQ920491.1  VCDJ01010665.1   95.318    1132        18        5       1   \n",
       "3  JQ920574.1  VCDJ01010665.1   98.889    3420         8        3       1   \n",
       "4  JQ920575.1  VCDJ01010665.1   97.664    2226        13        9    1789   \n",
       "5  EF990447.1  VCDJ01010675.1  100.000    2225         0        0     756   \n",
       "6  JQ920572.1  VCDJ01010675.1   98.425    2920        25        6       1   \n",
       "7  JQ920573.1  VCDJ01010675.1   99.604    4541        16        2    4072   \n",
       "8  JQ920570.1  VCDJ01010674.1   99.515     824         4        0       1   \n",
       "9  MH059954.1  VCDJ01010672.1   99.541   19606        62        7   59707   \n",
       "\n",
       "    qend    qlen    sstart      send      slen  evalue  bitscore  qcovs  \\\n",
       "0   1101    1101  19582530  19583630  54415314     0.0      2028    100   \n",
       "1   1679    1679  19570565  19572208  54415314     0.0      2857    100   \n",
       "2   1128    1128  19582530  19583630  54415314     0.0      1764    100   \n",
       "3   3392    3392  19567103  19570520  54415314     0.0      6080    100   \n",
       "4   4014    7419  19581984  19584170  54415314     0.0      3786     30   \n",
       "5   2980    3839  30228279  30230503  51265060     0.0      4109     58   \n",
       "6   2901    2901  30224409  30227326  51265060     0.0      5118    100   \n",
       "7   8611   13521  30218214  30222753  51265060     0.0      8285     34   \n",
       "8    824     824  11421697  11422520  65212296     0.0      1500    100   \n",
       "9  79297  197905  15981373  15961781  62546397     0.0     35682     10   \n",
       "\n",
       "   qcovhsp  \n",
       "0      100  \n",
       "1      100  \n",
       "2      100  \n",
       "3      100  \n",
       "4       30  \n",
       "5       58  \n",
       "6      100  \n",
       "7       34  \n",
       "8      100  \n",
       "9       10  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "blast_results = pd.read_table(\"../HCN_griffithsGenome_Blast.txt\", sep = '\\t')\n",
    "blast_results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 3: Retrieve contigs with matches to HCN genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "blast_results = '/scratch/research/projects/trifolium/HCN_genome_BLAST/HCN_griffithsGenome_Blast.txt'\n",
    "genome_fasta = '/scratch/research/references/trifolium/repens/griffiths_2019_Plant-cell/VCDJ01.1.fsa_nt'\n",
    "outfile = 'HCN_griffithsGenome_matchedScaffolds.fasta'\n",
    "retrieve_contigs(blast_results, genome_fasta, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
